cranelift-codegen/meta/src/shared/legalize.rs

```diff
@@ -272,6 +293,60 @@ pub(crate) fn define(insts: &InstructionGroup, imm: &Immediates) -> TransformGro
         ],
     );
 
+    // TODO(ryzokuken): benchmark this and decide if branching is a faster
+    // approach than evaluating boolean expressions.
+
+    narrow.custom_legalize(icmp_imm, "narrow_icmp_imm");
+
+    let intcc_eq = Literal::enumerator_for(&imm.intcc, "eq");
+    let intcc_ne = Literal::enumerator_for(&imm.intcc, "ne");
+    for &(int_ty, int_ty_half) in &[(I64, I32), (I128, I64)] {
+        narrow.legalize(
+            def!(b = icmp.int_ty(intcc_eq, x, y)),
+            vec![
+                def!((xl, xh) = isplit(x)),
+                def!((yl, yh) = isplit(y)),
+                def!(b1 = icmp.int_ty_half(intcc_eq, xl, yl)),
+                def!(b2 = icmp.int_ty_half(intcc_eq, xh, yh)),
+                def!(b = band(b1, b2)),
+            ],
+        );
+
+        narrow.legalize(
+            def!(b = icmp.int_ty(intcc_ne, x, y)),
+            vec![
+                def!((xl, xh) = isplit(x)),
+                def!((yl, yh) = isplit(y)),
+                def!(b1 = icmp.int_ty_half(intcc_ne, xl, yl)),
+                def!(b2 = icmp.int_ty_half(intcc_ne, xh, yh)),
+                def!(b = bor(b1, b2)),
+            ],
+        );
+
+        for cc in &["sgt", "sge", "slt", "sle", "ugt", "uge", "ult", "ule"] {
+            let intcc_cc = Literal::enumerator_for(&imm.intcc, cc);
+            let cc1 = Literal::enumerator_for(&imm.intcc, without_equal(cc));
+            let cc2 = Literal::enumerator_for(&imm.intcc, inverse_without_equal(cc));
+            let cc3 = Literal::enumerator_for(&imm.intcc, unsigned(cc));
+            narrow.legalize(
+                def!(b = icmp.int_ty(intcc_cc, x, y)),
+                vec![
+                    def!((xl, xh) = isplit(x)),
+                    def!((yl, yh) = isplit(y)),
+                    // X = A || (!A && !B && C)
+                    def!(b1 = icmp.int_ty_half(cc1, xh, yh)),
+                    def!(b2 = icmp.int_ty_half(cc2, xh, yh)),
+                    def!(b3 = icmp.int_ty_half(cc3, xl, yl)),
+                    def!(c1 = bnot(b1)),
+                    def!(c2 = bnot(b2)),
+                    def!(d1 = band(c1, c2)),
+                    def!(d2 = band(d1, b3)),
+                    def!(b = bor(b1, d2)),
```

Comment:
In this case, there are so many comparisons that I expect branching to be faster...

@sunfishcode, @abrown, you may know about this: in general, which is faster: materializing the flags register and grouping results of several comparisons with binary OR/AND, or using branches to short-circuit unnecessary comparisons?

Reply:
Not sure...